docs/topics/using-builders-with-builder-inference.md

```diff
@@ -0,0 +1,253 @@
+  ```
+* Inferring type arguments of several builder lambdas within one call including interdependent ones
+  ```kotlin
+  fun <K, V> myBuilder(
+      listBuilder: MutableList<V>.() -> Unit,
+      mapBuilder: MutableMap<K, V>.() -> Unit
+  ): Pair<List<V>, Map<K, V>> =
+      mutableListOf<V>().apply(listBuilder) to mutableMapOf<K, V>().apply(mapBuilder)
+  
+  fun main() {
+      val result = myBuilder(
+          { add(1) },
+          { put("key", 2) }
+      )
+      // result has Pair<List<Int>, Map<String, Int>> type
+  }
+  ```
+* Inferring type arguments whose type parameters are lambda's parameter or return types
+  ```kotlin
+  fun <K, V> myBuilder1(
+      mapBuilder: MutableMap<K, V>.() -> K
+  ): Map<K, V> = mutableMapOf<K, V>().apply { mapBuilder() }
+  
+  fun <K, V> myBuilder2(
+      mapBuilder: MutableMap<K, V>.(K) -> Unit
+  ): Map<K, V> = mutableMapOf<K, V>().apply { mapBuilder(2 as K) }
+  
+  fun main() {
+      // result1 has the Map<Long, String> type inferred
+      val result1 = myBuilder1 {
+          put(1L, "value")
+          2
+      }
+      val result2 = myBuilder2 {
+          put(1, "value 1")
+          // You can use `it` as "postponed type variable" type
+          // See the details in the section below
+          put(it, "value 2")
+      }
+  }
+  ```
+
+## How builder inference works
+
+### Postponed type variables
+
+Builder inference works in terms of _postponed type variables_, which appear inside builder lambda during builder 
+inference analysis. Postponed type variable is a type argument's type, which is in the process of inferring. 
+This means that the compiler still doesn't associate it with some concrete type but uses it for collecting a type 
+information about the type argument.
+
+Consider the example with [`buildList()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/build-list.html):
+
+```kotlin
+val result = buildList {
+    val x = get(0)
+}
+```
+
+Here `x` has a type of postponed type variable: the `get()` call returns a value of `E` type but `E` itself is not yet 
+fixed. At this moment, a concrete type for `E` is unknown. 
+
+When a value of postponed type variable gets associated with some concrete type, builder inference uses this association 
+to infer the resulting type of the corresponding type argument. For example:
+
+```kotlin
+val result = buildList {
+    val x = get(0)
+    val y: String = x
+} // result has the List<String> type inferred
+```
+
+After the postponed type variable gets assigned to a variable of the `String` type, builder inference can infer `E` into
+`String`.
+
+Note that you can always call `equals()`, `hashCode()`, and `toString()` functions with a postponed type variable as a 
+receiver. Once the postponed type variable gets associated with some concrete type, you can use it as a call argument, 
+call available extensions, and access available properties.
```

Comment:
> Once the postponed type variable gets associated with some concrete type, you can use it as a call argument, 
call available extensions, and access available properties.

Haven't got it. Inside a builder lambda, we have _always_ not fixed yet type variable (postponed type variable). What is the difference between the state "associated with some concrete type" and "not associated yet"? Associating is an adding constraints technically, but not variable fixation. We fix postponed variables only after _full_ analysis of a builder lambda. States "added some constraints" and "there aren't any added constraints yet" actually the same.

Reply:
Thank you, I indeed got it wrong. Changed the wording in multiple places of this section. [Please take a look at the last commit with review fixes.](https://github.com/JetBrains/kotlin-web-site/pull/2621/commits/f4e87dcde0888c7a1cfff7f69a22dc4cb13ed26e) I removed the sentence you cited here because now I get that it's about contributing type information, which we describe in the next section.